File: apps/android/app/src/main/java/app/inkuna/android/debug/ParityDigestRunner.kt
package app.inkuna.android.debug

import android.content.Context
import android.util.Log
import app.inkuna.android.model.LibraryStore
import app.inkuna.core.ImportOutcome
import app.inkuna.core.InkunaException
import app.inkuna.core.LayoutListener
import app.inkuna.core.ReaderLayoutSettings
import app.inkuna.core.Viewport
import java.io.File
import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import org.json.JSONArray
import org.json.JSONObject

/** Debug-only cross-shell layout-digest harness driven entirely by its manifest. */
object ParityDigestRunner {
    private const val TAG = "InkunaParity"
    private const val BOOK_TIMEOUT_MS = 120_000L

    suspend fun run(context: Context) = withContext(Dispatchers.Default) {
        val externalFiles = context.getExternalFilesDir(null)
        if (externalFiles == null) {
            Log.e(TAG, "PARITY ERROR External files directory unavailable")
            return@withContext
        }

        val corpus = File(externalFiles, "ParityCorpus")
        val manifest = File(corpus, "manifest.json")
        try {
            val cases = parseManifest(manifest)
            val output = cases.map { item ->
                File(item.file).name to digestWithTimeout(context.applicationContext, corpus, item)
            }
            File(externalFiles, "parity-android.json").writeText(encodeOutput(output))
            Log.i(TAG, "PARITY DONE ${cases.size}")
        } catch (failure: Throwable) {
            Log.e(TAG, "PARITY ERROR ${errorMessage(failure)}", failure)
        }
    }

    private suspend fun digestWithTimeout(
        context: Context,
        corpus: File,
        item: ParityManifestCase,
    ): ParityBookValue {
        val result = CompletableDeferred<ParityBookValue>()
        val scope = CoroutineScope(SupervisorJob() + Dispatchers.Default)
        scope.launch {
            result.complete(digest(context, corpus, item))
        }
        scope.launch {
            delay(BOOK_TIMEOUT_MS)
            result.complete(ParityBookValue.StringValue("TIMEOUT"))
        }
        return try {
            result.await()
        } finally {
            scope.cancel()
        }
    }

    private suspend fun digest(
        context: Context,
        corpus: File,
        item: ParityManifestCase,
    ): ParityBookValue {
        val book = File(corpus, item.file)
        if (!book.isFile) return ParityBookValue.StringValue("ERROR: File not found: ${item.file}")

        return try {
            val shelf = LibraryStore.bookshelf(context)
            val imported = shelf.importer().`import`(book.absolutePath)
            val publicationId = when (imported) {
                is ImportOutcome.Imported -> imported.publication.id
                is ImportOutcome.Duplicate -> imported.publication.id
                is ImportOutcome.Failed -> {
                    return ParityBookValue.StringValue("ERROR: ${errorMessage(imported.error)}")
                }
            }

            val completion = ParityCompletion()
            val session = shelf.openReader(
                publicationId,
                item.viewport.coreValue,
                item.settings.coreValue,
                ParityLayoutListener(completion),
            )
            try {
                val spineCount = session.spineCount()
                // `chapter` is cache-only: its expected NotReady result queues every
                // spine through the engine's own worker before we await terminals.
                repeat(spineCount.toInt()) { spine ->
                    runCatching { session.chapter(spine.toUInt()) }
                }
                val failedSpine = completion.waitForTerminal(spineCount)
                if (failedSpine != null) {
                    ParityBookValue.StringValue("ERROR: chapter $failedSpine failed")
                } else {
                    val chapters = mutableListOf<ParityChapterDigest>()
                    for (spine in 0 until spineCount.toInt()) {
                        val spineIdx = spine.toUInt()
                        while (true) {
                            val terminalEvents = completion.terminalEventCount(spineIdx)
                            try {
                                val chapter = session.chapter(spineIdx)
                                val pages = buildList {
                                    repeat(chapter.pageCount.toInt()) { page ->
                                        add(session.pageDigest(spineIdx, page.toUInt()))
                                    }
                                }
                                chapters += ParityChapterDigest(spineIdx, pages)
                                break
                            } catch (_: InkunaException.NotReady) {
                                if (completion.waitForTerminalEvent(spineIdx, terminalEvents)) {
                                    return ParityBookValue.StringValue("ERROR: chapter $spineIdx failed")
                                }
                            } catch (failure: Throwable) {
                                return ParityBookValue.StringValue("ERROR: ${errorMessage(failure)}")
                            }
                        }
                    }
                    ParityBookValue.Pages(chapters)
                }
            } finally {
                session.close()
            }
        } catch (failure: Throwable) {
            ParityBookValue.StringValue("ERROR: ${errorMessage(failure)}")
        }
    }

    private fun parseManifest(manifest: File): List<ParityManifestCase> {
        val source = JSONArray(manifest.readText())
        return List(source.length()) { index ->
            val item = source.getJSONObject(index)
            val viewport = item.requiredObject("viewport")
            val settings = item.requiredObject("settings")
            ParityManifestCase(
                file = item.requiredString("file"),
                viewport = ParityManifestViewport(
                    width = viewport.requiredPositiveFiniteDouble("width"),
                    height = viewport.requiredPositiveFiniteDouble("height"),
                ),
                settings = ParityManifestSettings(
                    readingFont = settings.requiredString("reading_font"),
                    readingBold = settings.requiredBoolean("reading_bold"),
                    textSizeStep = settings.requiredUByte("text_size_step"),
                    lineSpacing = settings.requiredFiniteDouble("line_spacing"),
                    letterSpacing = settings.requiredFiniteDouble("letter_spacing"),
                    wordSpacing = settings.requiredFiniteDouble("word_spacing"),
                    readingMargins = settings.requiredUInt("reading_margins"),
                ),
            )
        }
    }

    private fun encodeOutput(books: List<Pair<String, ParityBookValue>>): String = buildString {
        append('{')
        books.forEachIndexed { index, (file, value) ->
            if (index > 0) append(',')
            append(JSONObject.quote(file))
            append(':')
            append(value.jsonValue())
        }
        append('}')
    }

    private fun ParityBookValue.jsonValue(): String = when (this) {
        is ParityBookValue.StringValue -> JSONObject.quote(value)
        is ParityBookValue.Pages -> buildString {
            append('{')
            chapters.forEachIndexed { index, chapter ->
                if (index > 0) append(',')
                append(JSONObject.quote(chapter.spineIdx.toString()))
                append(':')
                append(JSONArray(chapter.digests).toString())
            }
            append('}')
        }
    }

    private fun errorMessage(error: Throwable): String = when (error) {
        is InkunaException.Io -> error.detail
        is InkunaException.FileTooLarge -> "limit=${error.limit}"
        is InkunaException.Database -> error.detail
        is InkunaException.Archive -> error.detail
        is InkunaException.UnsupportedFormat -> "format=${error.format ?: "nil"}"
        is InkunaException.InvalidPublication -> error.detail
        is InkunaException.NotReady -> error.detail
        is InkunaException.UnsupportedContent -> error.detail
        is InkunaException.LayoutBudgetExceeded -> error.detail
        is InkunaException.AnchorNotFound -> error.detail
        is InkunaException.Search -> error.detail
        is InkunaException.NotFound -> "id=${error.id}"
        else -> error.message ?: error.javaClass.simpleName
    }
}

private data class ParityManifestCase(
    val file: String,
    val viewport: ParityManifestViewport,
    val settings: ParityManifestSettings,
)

private data class ParityManifestViewport(
    val width: Double,
    val height: Double,
) {
    val coreValue: Viewport
        get() = Viewport(width, height)
}

private data class ParityManifestSettings(
    val readingFont: String,
    val readingBold: Boolean,
    val textSizeStep: UByte,
    val lineSpacing: Double,
    val letterSpacing: Double,
    val wordSpacing: Double,
    val readingMargins: UInt,
) {
    val coreValue: ReaderLayoutSettings
        get() = ReaderLayoutSettings(
            readingFont,
            readingBold,
            textSizeStep,
            lineSpacing,
            letterSpacing,
            wordSpacing,
            readingMargins,
        )
}

private data class ParityChapterDigest(
    val spineIdx: UInt,
    val digests: List<String>,
)

private sealed interface ParityBookValue {
    data class Pages(val chapters: List<ParityChapterDigest>) : ParityBookValue

    data class StringValue(val value: String) : ParityBookValue
}

private fun JSONObject.requiredObject(name: String): JSONObject = requiredValue(name) as? JSONObject
    ?: throw IllegalArgumentException("$name must be an object")

private fun JSONObject.requiredString(name: String): String = requiredValue(name) as? String
    ?: throw IllegalArgumentException("$name must be a string")

private fun JSONObject.requiredBoolean(name: String): Boolean = requiredValue(name) as? Boolean
    ?: throw IllegalArgumentException("$name must be a boolean")

private fun JSONObject.requiredPositiveFiniteDouble(name: String): Double {
    val value = requiredFiniteDouble(name)
    require(value > 0) { "$name must be positive" }
    return value
}

private fun JSONObject.requiredFiniteDouble(name: String): Double {
    val value = requiredValue(name) as? Number
        ?: throw IllegalArgumentException("$name must be a number")
    val number = value.toDouble()
    require(number.isFinite()) { "$name must be finite" }
    return number
}

private fun JSONObject.requiredUByte(name: String): UByte {
    val value = requiredInteger(name)
    require(value in UByte.MIN_VALUE.toLong()..UByte.MAX_VALUE.toLong()) { "$name is out of range" }
    return value.toUByte()
}

private fun JSONObject.requiredUInt(name: String): UInt {
    val value = requiredInteger(name)
    require(value in 0L..UInt.MAX_VALUE.toLong()) { "$name is out of range" }
    return value.toUInt()
}

private fun JSONObject.requiredInteger(name: String): Long = when (val value = requiredValue(name)) {
    is Int -> value.toLong()
    is Long -> value
    else -> throw IllegalArgumentException("$name must be an integer")
}

private fun JSONObject.requiredValue(name: String): Any {
    val value = get(name)
    require(value !== JSONObject.NULL) { "$name must not be null" }
    return value
}

private class ParityCompletion {
    private val terminal = linkedSetOf<UInt>()
    private val terminalEventCounts = mutableMapOf<UInt, Int>()
    private val terminalEventFailures = mutableMapOf<UInt, Boolean>()
    private val terminalEventWaiters = mutableMapOf<UInt, CompletableDeferred<Boolean>>()
    private var failedSpine: UInt? = null
    private var expected: UInt? = null
    private val completion = CompletableDeferred<UInt?>()

    fun chapterReady(spineIdx: UInt) = record(spineIdx, failed = false)

    fun chapterFailed(spineIdx: UInt) = record(spineIdx, failed = true)

    suspend fun waitForTerminal(spineCount: UInt): UInt? {
        synchronized(this) {
            expected = spineCount
            completeIfTerminal()
        }
        return completion.await()
    }

    fun terminalEventCount(spineIdx: UInt): Int = synchronized(this) {
        terminalEventCounts[spineIdx] ?: 0
    }

    suspend fun waitForTerminalEvent(spineIdx: UInt, after: Int): Boolean {
        val waiter = CompletableDeferred<Boolean>()
        var immediate: Boolean? = null
        synchronized(this) {
            if ((terminalEventCounts[spineIdx] ?: 0) > after) {
                immediate = terminalEventFailures[spineIdx] ?: false
            } else {
                terminalEventWaiters[spineIdx] = waiter
            }
        }
        return immediate ?: waiter.await()
    }

    private fun record(spineIdx: UInt, failed: Boolean) {
        synchronized(this) {
            terminalEventCounts[spineIdx] = (terminalEventCounts[spineIdx] ?: 0) + 1
            terminalEventFailures[spineIdx] = failed
            terminalEventWaiters.remove(spineIdx)?.complete(failed)
            if (!terminal.add(spineIdx)) return
            if (failed) failedSpine = minOf(failedSpine ?: spineIdx, spineIdx)
            completeIfTerminal()
        }
    }

    private fun completeIfTerminal() {
        val expected = expected ?: return
        if (terminal.size >= expected.toInt()) completion.complete(failedSpine)
    }
}

private class ParityLayoutListener(
    private val completion: ParityCompletion,
) : LayoutListener {
    override fun onFirstPageReady(generation: ULong, spineIdx: UInt) = Unit

    override fun onChapterReady(generation: ULong, spineIdx: UInt, pageCount: UInt) {
        completion.chapterReady(spineIdx)
    }

    override fun onChapterFailed(generation: ULong, spineIdx: UInt) {
        completion.chapterFailed(spineIdx)
    }
}

File: apps/ios/Inkuna/Debug/ParityDigestRunner.swift
#if DEBUG
import Foundation

/// Scriptable cross-shell layout-digest harness. The manifest carries every
/// layout value so this runner cannot accidentally become a second settings
/// source of truth.
enum ParityDigestRunner {
    @MainActor
    static func run() async {
        guard let documents = FileManager.default.urls(
            for: .documentDirectory,
            in: .userDomainMask
        ).first else {
            print("PARITY ERROR Documents directory unavailable")
            return
        }

        let corpus = documents.appendingPathComponent("ParityCorpus", isDirectory: true)
        let manifestURL = corpus.appendingPathComponent("manifest.json")
        do {
            let manifest = try JSONDecoder().decode(
                [ParityManifestCase].self,
                from: Data(contentsOf: manifestURL)
            )
            for item in manifest {
                try item.validate()
            }
            var output = ParityDigestOutput()
            for item in manifest {
                output.append(
                    file: URL(fileURLWithPath: item.file).lastPathComponent,
                    value: await digestWithTimeout(item, corpus: corpus)
                )
            }
            try output.encodedData().write(
                to: documents.appendingPathComponent("parity-ios.json"),
                options: .atomic
            )
            print("PARITY DONE \(manifest.count) books")
        } catch {
            print("PARITY ERROR \(errorMessage(error))")
        }
    }

    private static func digestWithTimeout(
        _ item: ParityManifestCase,
        corpus: URL
    ) async -> ParityBookValue {
        let gate = ParityResultGate()
        let work = Task {
            await gate.resolve(await digest(item, corpus: corpus))
        }
        let timeout = Task {
            try? await Task.sleep(for: .seconds(120))
            await gate.resolve(.string("TIMEOUT"))
        }
        let value = await gate.wait()
        work.cancel()
        timeout.cancel()
        return value
    }

    private static func digest(_ item: ParityManifestCase, corpus: URL) async -> ParityBookValue {
        let bookURL = corpus.appendingPathComponent(item.file)
        guard FileManager.default.fileExists(atPath: bookURL.path) else {
            return .string("ERROR: File not found: \(item.file)")
        }

        do {
            let shelf = try await LibraryStore.shared.library()
            let imported = try await shelf.importer().import(path: bookURL.path)
            let publicationID: String
            switch imported {
            case let .imported(publication), let .duplicate(publication):
                publicationID = publication.id
            case let .failed(_, error):
                return .string("ERROR: \(errorMessage(error))")
            }

            let completion = ParityCompletion()
            let listener = ParityLayoutListener(completion: completion)
            let session = try await shelf.openReader(
                id: publicationID,
                viewport: item.viewport.coreValue,
                settings: item.settings.coreValue,
                listener: listener
            )
            let spineCount = session.spineCount()
            // `chapter` is cache-only: its expected NotReady result queues every
            // spine through the engine's own worker before we await terminals.
            for spineIdx in 0..<spineCount {
                _ = try? session.chapter(spineIdx: spineIdx)
            }
            if let failedSpine = try await completion.wait(for: spineCount) {
                return .string("ERROR: chapter \(failedSpine) failed")
            }

            var chapters: [ParityChapterDigest] = []
            for spineIdx in 0..<spineCount {
                while true {
                    let terminalEvents = await completion.terminalEventCount(for: spineIdx)
                    do {
                        let chapter = try session.chapter(spineIdx: spineIdx)
                        var pages: [String] = []
                        for pageIdx in 0..<chapter.pageCount {
                            pages.append(try session.pageDigest(spineIdx: spineIdx, pageIdx: pageIdx))
                        }
                        chapters.append(ParityChapterDigest(spineIdx: spineIdx, digests: pages))
                        break
                    } catch let error as InkunaError {
                        guard case .NotReady = error else {
                            return .string("ERROR: \(errorMessage(error))")
                        }
                        if try await completion.waitForTerminalEvent(spineIdx, after: terminalEvents) {
                            return .string("ERROR: chapter \(spineIdx) failed")
                        }
                    } catch {
                        return .string("ERROR: \(errorMessage(error))")
                    }
                }
            }
            return .pages(chapters)
        } catch {
            return .string("ERROR: \(errorMessage(error))")
        }
    }

    private static func errorMessage(_ error: Error) -> String {
        guard let error = error as? InkunaError else {
            return error.localizedDescription
        }
        switch error {
        case let .Io(detail), let .Database(detail), let .Archive(detail),
             let .InvalidPublication(detail), let .NotReady(detail),
             let .UnsupportedContent(detail), let .LayoutBudgetExceeded(detail),
             let .AnchorNotFound(detail), let .Search(detail):
            return detail
        case let .FileTooLarge(limit):
            return "limit=\(limit)"
        case let .UnsupportedFormat(format):
            return "format=\(format ?? "nil")"
        case let .NotFound(id):
            return "id=\(id)"
        }
    }
}

private struct ParityManifestCase: Decodable, Sendable {
    let file: String
    let viewport: ParityManifestViewport
    let settings: ParityManifestSettings

    func validate() throws {
        guard viewport.width.isFinite, viewport.width > 0, viewport.height.isFinite, viewport.height > 0 else {
            throw ParityManifestError.invalidViewport(file)
        }
    }
}

private enum ParityManifestError: LocalizedError {
    case invalidViewport(String)

    var errorDescription: String? {
        switch self {
        case let .invalidViewport(file):
            return "Invalid viewport for \(file)"
        }
    }
}

private struct ParityManifestViewport: Decodable, Sendable {
    let width: Double
    let height: Double

    var coreValue: Viewport {
        Viewport(width: width, height: height)
    }
}

private struct ParityManifestSettings: Decodable, Sendable {
    let readingFont: String
    let readingBold: Bool
    let textSizeStep: UInt8
    let lineSpacing: Double
    let letterSpacing: Double
    let wordSpacing: Double
    let readingMargins: UInt32

    private enum CodingKeys: String, CodingKey {
        case readingFont = "reading_font"
        case readingBold = "reading_bold"
        case textSizeStep = "text_size_step"
        case lineSpacing = "line_spacing"
        case letterSpacing = "letter_spacing"
        case wordSpacing = "word_spacing"
        case readingMargins = "reading_margins"
    }

    var coreValue: ReaderLayoutSettings {
        ReaderLayoutSettings(
            readingFont: readingFont,
            readingBold: readingBold,
            textSizeStep: textSizeStep,
            lineSpacing: lineSpacing,
            letterSpacing: letterSpacing,
            wordSpacing: wordSpacing,
            readingMargins: readingMargins
        )
    }
}

private struct ParityChapterDigest: Sendable {
    let spineIdx: UInt32
    let digests: [String]
}

private enum ParityBookValue: Sendable {
    case pages([ParityChapterDigest])
    case string(String)
}

private struct ParityDigestOutput {
    private var books: [(String, ParityBookValue)] = []

    mutating func append(file: String, value: ParityBookValue) {
        books.append((file, value))
    }

    func encodedData() throws -> Data {
        let entries = try books.map { file, value in
            "\(try jsonString(file)):\(try jsonValue(value))"
        }
        return Data("{\(entries.joined(separator: ","))}".utf8)
    }

    private func jsonValue(_ value: ParityBookValue) throws -> String {
        switch value {
        case let .string(message):
            return try jsonString(message)
        case let .pages(chapters):
            let entries = try chapters.map { chapter in
                "\(try jsonString(String(chapter.spineIdx))):\(try jsonArray(chapter.digests))"
            }
            return "{\(entries.joined(separator: ","))}"
        }
    }

    private func jsonArray(_ values: [String]) throws -> String {
        "[\(try values.map(jsonString).joined(separator: ","))]"
    }

    private func jsonString(_ value: String) throws -> String {
        String(decoding: try JSONEncoder().encode(value), as: UTF8.self)
    }
}

private actor ParityResultGate {
    private var value: ParityBookValue?
    private var waiter: CheckedContinuation<ParityBookValue, Never>?

    func resolve(_ newValue: ParityBookValue) {
        guard value == nil else { return }
        value = newValue
        waiter?.resume(returning: newValue)
        waiter = nil
    }

    func wait() async -> ParityBookValue {
        if let value { return value }
        return await withCheckedContinuation { waiter = $0 }
    }
}

private actor ParityCompletion {
    private var terminalSpines = Set<UInt32>()
    private var terminalEventCounts: [UInt32: Int] = [:]
    private var terminalEventFailures: [UInt32: Bool] = [:]
    private var terminalEventWaiters: [UInt32: CheckedContinuation<Bool, Error>] = [:]
    private var failedSpine: UInt32?
    private var expectedSpines: UInt32?
    private var waiter: CheckedContinuation<UInt32?, Error>?

    func chapterReady(_ spineIdx: UInt32) {
        record(spineIdx, failed: false)
    }

    func chapterFailed(_ spineIdx: UInt32) {
        record(spineIdx, failed: true)
    }

    func wait(for spineCount: UInt32) async throws -> UInt32? {
        expectedSpines = spineCount
        if terminalSpines.count >= Int(spineCount) { return failedSpine }
        return try await withTaskCancellationHandler(operation: {
            try await withCheckedThrowingContinuation { continuation in
                if Task.isCancelled {
                    continuation.resume(throwing: CancellationError())
                } else {
                    waiter = continuation
                }
            }
        }, onCancel: {
            Task { await self.cancelInitialWait() }
        })
    }

    func terminalEventCount(for spineIdx: UInt32) -> Int {
        terminalEventCounts[spineIdx, default: 0]
    }

    func waitForTerminalEvent(_ spineIdx: UInt32, after count: Int) async throws -> Bool {
        if terminalEventCounts[spineIdx, default: 0] > count {
            return terminalEventFailures[spineIdx, default: false]
        }
        return try await withTaskCancellationHandler(operation: {
            try await withCheckedThrowingContinuation { continuation in
                if Task.isCancelled {
                    continuation.resume(throwing: CancellationError())
                } else {
                    terminalEventWaiters[spineIdx] = continuation
                }
            }
        }, onCancel: {
            Task { await self.cancelTerminalEventWait(for: spineIdx) }
        })
    }

    private func cancelInitialWait() {
        guard let waiter else { return }
        self.waiter = nil
        waiter.resume(throwing: CancellationError())
    }

    private func cancelTerminalEventWait(for spineIdx: UInt32) {
        guard let waiter = terminalEventWaiters.removeValue(forKey: spineIdx) else { return }
        waiter.resume(throwing: CancellationError())
    }

    private func record(_ spineIdx: UInt32, failed: Bool) {
        terminalEventCounts[spineIdx, default: 0] += 1
        terminalEventFailures[spineIdx] = failed
        let eventWaiter = terminalEventWaiters.removeValue(forKey: spineIdx)
        eventWaiter?.resume(returning: failed)
        guard terminalSpines.insert(spineIdx).inserted else { return }
        if failed { failedSpine = min(failedSpine ?? spineIdx, spineIdx) }
        guard
            let expectedSpines,
            terminalSpines.count >= Int(expectedSpines),
            let waiter
        else { return }
        self.waiter = nil
        waiter.resume(returning: failedSpine)
    }
}

private final class ParityLayoutListener: LayoutListener, @unchecked Sendable {
    private let completion: ParityCompletion

    init(completion: ParityCompletion) {
        self.completion = completion
    }

    func onFirstPageReady(generation _: UInt64, spineIdx _: UInt32) {}

    func onChapterReady(generation _: UInt64, spineIdx: UInt32, pageCount _: UInt32) {
        Task { await completion.chapterReady(spineIdx) }
    }

    func onChapterFailed(generation _: UInt64, spineIdx: UInt32) {
        Task { await completion.chapterFailed(spineIdx) }
    }
}
#endif
